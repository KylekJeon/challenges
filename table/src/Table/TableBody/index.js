import React from 'react';
import TableRow from '../TableRow';

// rows is an array of objects
const TableBody = rows => {
  return (
    <tbody>
      {
        rows.map(row => {
          return <TableRow row={row} />
        })
      }
    </tbody>
  )
}

export default TableBody;