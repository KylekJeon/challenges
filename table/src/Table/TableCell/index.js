import React from 'react';

const TableCell = ({data}) => {
  return (
    <td>
      {data}
    </td>
  )
}

export default TableCell;