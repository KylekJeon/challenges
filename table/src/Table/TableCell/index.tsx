import React, { StatelessComponent, Props } from 'react';

interface TableCellProps extends Props<void> {
  data: string;
}

const TableCell: StatelessComponent<TableCellProps> = function TableCell({ data }) {
  return (
    <td className="row">
      {data}
    </td>
  )
}

export default TableCell;