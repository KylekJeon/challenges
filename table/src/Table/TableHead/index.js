import React from 'react';

const TableHead = ({name, onHeaderClick}) => {
  const thOnClick = e => {
    onHeaderClick(e, name);
  }
  return (
    <th onClick={thOnClick}>
      {name.split('_').join(' ')}
    </th>
  )
}

export default TableHead;