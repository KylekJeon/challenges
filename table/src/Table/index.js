import React from 'react'
import styled from 'styled-components';
import Loader from 'react-loader-spinner';

const PageContainer = styled.div`
  width: 100%;
`

const LoaderContainer = styled.div`
  width: 100%;
  min-height: 70vh;
  display: flex;
  justify-content: center;
  align-items: center;
`

const Container = styled.div`
  border: 1px solid black;
  background: gray;
  padding: 16px;
`
const ColumnHeader = styled.div`
  font-weight: bold;
`

const ColumnHeaderContainer = styled.div`

`

class Table extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      pageIdx: 0,
      dataReady: false,
      nextUrl: null,
      dataArray: [],
      columnNames: [],
      activeSort: null
    };
  }

  componentDidMount(){
    fetch("https://swapi.co/api/people/")
    .then(res => res.json())
    .then(parsedRes => {
      // If table should be truly generic (no pre-existing awareness of fields whatsoever), data should be left unfiltered
      const dataArray = this.filterData(parsedRes.results);
      // create keys array to serve as column names
      const columnNames = Object.keys(dataArray[0]);

      this.setState({
        dataArray,
        columnNames,
        nextUrl: parsedRes.next,
        activeSort: columnNames[0]
      })
    })
    .catch(err => {
      console.log(err);
    })
  }

  filterData = (dataArray) => {
    const returnArray = [];
    
    // Select the specific data/columns that you want to display on table using object destructuring and property shorthand
    dataArray.forEach(data => {
      const filteredData = (({ name, height, mass, hair_color, eye_color, created, edited }) => ({ name, height, mass, hair_color, eye_color, created, edited }))(data);

      returnArray.push(filteredData);
    })

    return returnArray;
  }

  generateColumnNamesRow = () => {
    const columnHeaders = this.columnNames.map(name => (
      <ColumnHeader
        onClick={(name) => this.sortByColumn(name)}
      >
        {name}
      </ColumnHeader>
    ))

    return (
      <ColumnHeaderContainer>
        {columnHeaders}
      </ColumnHeaderContainer>
    )
  }

  sortByColumn = (name) => {
    this.setState({
      
    })
  }

  render() {
    let mainContent = (
      <LoaderContainer>
        <Loader 
          type="Puff"
          color="#00BFFF"
          height="200"	
          width="200"
        />   
      </LoaderContainer>
    )

    if(this.state.dataReady){
      const columnNamesRow = this.generateColumnNamesRow();
      
      mainContent = (
        <Container>
          {columnNamesRow}
        </Container>
      )
    }

    return (
      <PageContainer>
        {mainContent}
      </PageContainer>
    );
  }
}

export default Table;
