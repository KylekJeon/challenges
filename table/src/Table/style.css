.table {
  width: 100vh;
  margin: auto;
  border-collapse: collapse;
  text-align: left;
  color: #455576;
}