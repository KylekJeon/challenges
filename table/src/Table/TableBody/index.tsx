import React, { StatelessComponent, Props } from 'react';
import { Row } from '../../types/table'; 
import TableRow from '../TableRow';
import TableCell from '../TableCell';

interface TableBodyProps extends Props<void> {
  rows: Row[];
}

const TableBody: StatelessComponent<TableBodyProps> = function TableBody({
  rows,
}) {
  return (
    <tbody>
      {
        rows.map((row, i) => {
          return (
            <TableRow key={i}>
              {
                Object.keys(row).map((colTitle, i) => {
                  const data = row[colTitle];
                  return (
                    <TableCell key={i} data={data} />
                  )
                })
              }
            </TableRow>
          )
        })
      }
    </tbody>
  )
}

export default TableBody;