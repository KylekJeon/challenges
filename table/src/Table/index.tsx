import React from 'react'
import camelize from 'camelize';
import './style.css';
import TableHeader from './TableHeader';
import TableBody from './TableBody';
import { pick, isNull } from 'lodash';

export default class Table extends React.Component {
  state = {
    rows: null,
    columns: null,
    validHeaders: ['name', 'height', 'hairColor', 'eyeColor', 'created', 'edited'],
    isAscending: true,
  }

  fetchData = () => {
    return fetch('/api/people');
  }

  processFetchData = (data) => {
    const results = data.results;
    return results.map(result => {
      result.created = new Date(result.created).toLocaleDateString();
      result.edited = new Date(result.edited).toLocaleDateString();
      return pick(result, this.state.validHeaders);
    });
  }

  componentDidMount() {
    // fetch data if rows and cols are not there
    if (!this.state.rows && !this.state.columns) {
      this.fetchData()
      .then(resp => resp.json())
      .then(data => camelize(data))
      .then(data => {
        const processedData = this.processFetchData(data);
        const columns  = Object.keys(processedData[0]);
        const rows = processedData;
        
        this.setState({
          columns,
          rows,
        })
      });
    }
  }

  onHeaderClick = (colName: string) => {
    // sort by the colName
    this.setState((prev: any) => {
      return {
        isAscending: !prev.isAscending, 
        rows: prev.rows.sort((a, b) => {
            // ignore the case when compairing
            const stringA = a[colName].toString().toUpperCase();
            const stringB = b[colName].toString().toUpperCase();
            const toggle = prev.isAscending ? 1 : -1;
            
            if (stringA < stringB) {
              return -1 * toggle;
            }
            if (stringA > stringB) {
              return 1 * toggle;
            }
            return 0;
        }),
      }
    });
  }

  render() {
    const {
      columns,
      rows,
    } = this.state;
    const hasData = !isNull(rows) && !isNull(columns);
    return hasData ? (
      <table className="table" >
        <TableHeader columns={columns} onHeaderClick={this.onHeaderClick} />
        <TableBody rows={rows} />
      </table>
    ) : (
      <div className="loading">Loading...</div>
    );
  }
}
