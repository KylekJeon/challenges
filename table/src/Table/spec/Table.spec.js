import React from "react";
import { shallow } from "enzyme";
import Table from "..";
import camelize from "camelize";
import tableData from "../../test-data/first-page.json";
import { mockColumns, mockRows } from "../../test-data/mock-data";

describe("Table", () => {
  const testGlobal = global;
  const camelizedTableData = camelize(tableData);

  beforeEach(() => {
    testGlobal.fetch = jest.fn(() =>
      Promise.resolve({
        ok: true,
        url: "https://www.swapi.co/api/fake/people",
        json: () => camelizedTableData
      })
    );
  });

  afterEach(() => {
    testGlobal.fetch.mockClear();
  });

  it("should render loader with no data", () => {
    const component = shallow(<Table />);
    expect(component).toMatchSnapshot();
  });

  it("should render table with data", done => {
    const component = shallow(<Table />);

    process.nextTick(() => {
      component.update();
      expect(component).toMatchSnapshot();
      done();
    });
  });

  it("headerClick should update selected column", () => {
    const component = shallow(<Table />);
    component.setState({
      rows: mockRows,
      columns: mockColumns
    });

    component.instance().onHeaderClick(mockColumns[0]);
    expect(component.state("rows")[0].name).toBe("Beru Whitesun lars");
  });

  it("headerClick should toggle isAscending", () => {
    const component = shallow(<Table />);
    const selectedColumn = {
      ...mockColumns[0],
      selected: true,
      isAscending: true
    };
    const newColumns = [selectedColumn, ...mockColumns.slice(1)];
    component.setState({
      rows: mockRows,
      columns: newColumns
    });

    component.instance().onHeaderClick(newColumns[0]);
    expect(component.state("columns")[0].isAscending).toBe(false);
    expect(component.state("columns")[0].selected).toBe(true);
  });

  it("headerClick should sort dates", () => {
    const component = shallow(<Table />);
    component.setState({
      rows: mockRows,
      columns: mockColumns
    });

    component.instance().onHeaderClick(mockColumns[4]);
    expect(component.state("rows")[0].created).toBe("12/9/2014");
  });
});
