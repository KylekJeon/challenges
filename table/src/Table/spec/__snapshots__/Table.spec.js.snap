// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should render loader with no data 1`] = `
<div
  className="loading"
>
  Loading...
</div>
`;

exports[`Table should render table with data 1`] = `
<table
  className="table"
>
  <TableHeader
    columns={
      Array [
        Object {
          "isAscending": false,
          "name": "name",
          "selected": false,
        },
        Object {
          "isAscending": false,
          "name": "height",
          "selected": false,
        },
        Object {
          "isAscending": false,
          "name": "hairColor",
          "selected": false,
        },
        Object {
          "isAscending": false,
          "name": "eyeColor",
          "selected": false,
        },
        Object {
          "isAscending": false,
          "name": "created",
          "selected": false,
        },
        Object {
          "isAscending": false,
          "name": "edited",
          "selected": false,
        },
      ]
    }
    onHeaderClick={[Function]}
  />
  <TableBody
    rows={
      Array [
        Object {
          "created": "12/9/2014",
          "edited": "12/20/2014",
          "eyeColor": "blue",
          "hairColor": "blond",
          "height": 172,
          "name": "Luke Skywalker",
        },
        Object {
          "created": "12/10/2014",
          "edited": "12/20/2014",
          "eyeColor": "yellow",
          "hairColor": "n/a",
          "height": 167,
          "name": "C-3PO",
        },
        Object {
          "created": "12/10/2014",
          "edited": "12/20/2014",
          "eyeColor": "red",
          "hairColor": "n/a",
          "height": 96,
          "name": "R2-D2",
        },
        Object {
          "created": "12/10/2014",
          "edited": "12/20/2014",
          "eyeColor": "yellow",
          "hairColor": "none",
          "height": 202,
          "name": "Darth Vader",
        },
        Object {
          "created": "12/10/2014",
          "edited": "12/20/2014",
          "eyeColor": "brown",
          "hairColor": "brown",
          "height": 150,
          "name": "Leia Organa",
        },
        Object {
          "created": "12/10/2014",
          "edited": "12/20/2014",
          "eyeColor": "blue",
          "hairColor": "brown, grey",
          "height": 178,
          "name": "Owen Lars",
        },
        Object {
          "created": "12/10/2014",
          "edited": "12/20/2014",
          "eyeColor": "blue",
          "hairColor": "brown",
          "height": 165,
          "name": "Beru Whitesun lars",
        },
        Object {
          "created": "12/10/2014",
          "edited": "12/20/2014",
          "eyeColor": "red",
          "hairColor": "n/a",
          "height": 97,
          "name": "R5-D4",
        },
        Object {
          "created": "12/10/2014",
          "edited": "12/20/2014",
          "eyeColor": "brown",
          "hairColor": "black",
          "height": 183,
          "name": "Biggs Darklighter",
        },
        Object {
          "created": "12/10/2014",
          "edited": "12/20/2014",
          "eyeColor": "blue-gray",
          "hairColor": "auburn, white",
          "height": 182,
          "name": "Obi-Wan Kenobi",
        },
      ]
    }
  />
</table>
`;
